### ISSUE TITLE
Additional practice questions for part 2 cross check

### ISSUE THREAD

**Issue by kiwibang** [nus-cs2103-AY2324S1/forum#547]
Hi everyone, as per the title, I would like to cross check and discuss answers for the additional practice questions!

Here is my answers (apologies for the messy diagrams, was trying to finish within the 20 minute window):

**(a) and (b)**
![photo_2023-11-28 11 39 05](https://github.com/nus-cs2103-AY2324S1/forum/assets/66860275/dab78a67-0194-4044-80fa-73b20f527c4b)

**(c)**
![photo_2023-11-28 11 39 10](https://github.com/nus-cs2103-AY2324S1/forum/assets/66860275/0c36b958-5407-4e10-9f57-ef89bc178684)

**(d)**
![photo_2023-11-28 11 39 08](https://github.com/nus-cs2103-AY2324S1/forum/assets/66860275/83b01257-2f87-4420-8c76-0803d71d3df1)

**Comment by GohTengFong** [comment]
Q1(a) How did you know that `heading` is a package private visibility? Other than this, my diagram is similar to yours
Q1(b) Instead of 2 arrows, mine is drawn as composition (similar to 1(a)). I also added 2 more reference (dotted) arrows from each Section object to the Handout object
Q1(c) I did not have an [else] condition in my `alt` box. Other than that, my diagram is similar

**Comment by damithc** [comment]
@kiwibang thanks for sharing your answers (and @GohTengFong for inputs). I'll leave you all to discuss and figure out the correct answer. If there are still unresolved points after discussion, let me know so that I can pitch in.

**Comment by damithc** [comment]
On a related note, recommended to follow the given layout for the class diagram. It will help when you match your diagram with diagram choices given in exam part 3.

**Comment by jannnice** [comment]
I share the same question as @GohTengFong, how do we decide on when to put package private? Is it correct for me to say that any attributes without visibility will be considered as package private?
For example:
`String heading` -> ~heading:String
`public String heading` -> +heading:String

Another question is, do we only draw the composition relationship when the question gives the "explicit keywords"?

**Comment by damithc** [comment]
> Another question is, do we only draw the composition relationship when the question gives the "explicit keywords"?

@jannnice If composition is expected, there will be an additional hint to nudge you towards it. So, you can first figure out which associations are likely to be composition, and look for the hint to confirm.

**Comment by kiwibang** [comment]
> Q1(a) How did you know that `heading` is a package private visibility?


`heading` does not have any access modifier. In Java, any field without an access modifier is package-private by default as per the [official Java documentation](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

Here is a screenshot of the relevant portion of the `Handout` class for easier reference:

<img width="443" alt="Screenshot 2023-11-28 at 6 00 35 PM" src="https://github.com/nus-cs2103-AY2324S1/forum/assets/66860275/adcdf777-0627-43c5-9f9d-3761f0edaedb">


> Q1(b) Instead of 2 arrows, mine is drawn as composition (similar to 1(a)). I also added 2 more reference (dotted) arrows from each Section object to the Handout object 


I think I should have added those for more detail. Thanks for pointing that out!

> Q1(c) I did not have an [else] condition in my `alt` box. Other than that, my diagram is similar

I think omitting the `[else]` is acceptable in activity diagrams. Page 155 of the textbook does mention this:
<img width="495" alt="Screenshot 2023-11-28 at 5 59 10 PM" src="https://github.com/nus-cs2103-AY2324S1/forum/assets/66860275/74a27e41-b475-4af7-9b29-406c2e071e85">

**Comment by aexolate** [comment]
For 1d, will the 'place hand' at the [else] branch be extra?

e.g. Place Hand on Reader -> verification failed -[else]-> **Place hand -> Place Hand on Reader** -> ...

**Comment by TeeRenJing** [comment]
![20231128_215938.jpg](https://github.com/nus-cs2103-AY2324S1/forum/assets/62975067/841644a0-081d-4963-b189-61cd1a9c2486)

I was thinking of an activity diagram like this instead.

**Comment by Elijah5399** [comment]
For the activity diagram question, I agree with @TeeRenJing 's implementation of the if-else portion. I think it makes more sense to have 3 different cases: 
1. Where p1, p2 and p3 are all positive,
2. There are 3 consecutive failures, and
3. Else

However, I also added a box containing "user prompted to place hand again" to the else line. I feel like @kiwibang 's way of representing the if-else statements is inaccurate. According to his diagram, 3 failures causes the user to be prompted to place their hand again, followed by the door being disabled, which doesn't make sense.

In response to @GohTengFong 's comment, I believe including the dotted arrows in the object diagram is incorrect, as the `Handout` depended on by the `Section` object may not be the same `Handout` object which it composes. 

With regards to the sequence diagram, within the `[alt]` block, I believe an empty `[else]` block is necessary. To my understanding, inside the `[alt]` block, exactly one scenario has to be chosen.

**Comment by richiehx** [comment]
@Elijah5399 From the textbook, it is ok for none of the alt blocks to be taken, so I believe an empty else block is not necessary.

![image](https://github.com/nus-cs2103-AY2324S1/forum/assets/122219138/2fe9f433-1175-42c4-a72d-3cd0d579edd9)

**Comment by ktzy0305** [comment]
> Q1(a) How did you know that `heading` is a package private visibility? Other than this, my diagram is similar to yours Q1(b) Instead of 2 arrows, mine is drawn as composition (similar to 1(a)). I also added 2 more reference (dotted) arrows from each Section object to the Handout object Q1(c) I did not have an [else] condition in my `alt` box. Other than that, my diagram is similar

@GohTengFong I don't think there is a need to add the 2 reference arrows for the object diagram.

**Comment by ktzy0305** [comment]
Either way here are my answers

**(a)**
![photo_2023-11-30 11 15 17](https://github.com/nus-cs2103-AY2324S1/forum/assets/36191443/22fb472d-15a1-4099-a124-0066590cb510)

**(b)**
![photo_2023-11-30 11 15 16](https://github.com/nus-cs2103-AY2324S1/forum/assets/36191443/ce513ea9-ca15-4bdc-ba60-b7e4ac1a7544)


**(c)**
There is a mistake where I missed out the final partition [else] in the alt frame. Other than that its mostly the same as @kiwibang.
![photo_2023-11-30 11 15 14](https://github.com/nus-cs2103-AY2324S1/forum/assets/36191443/0e56b0f8-2ca8-4bce-aef9-e7a68210948b)


**(d)**
![photo_2023-11-30 11 15 11](https://github.com/nus-cs2103-AY2324S1/forum/assets/36191443/9f76c8c2-00db-4775-b72a-739265197c13)

**Comment by jianrong7** [comment]
I realise that the code in part c does not have `break` statements in the `switch-case` block. Will this affect the sequence diagram? Because if `status == 2`, the code will execute `f.increase(2)`, then it will fallthrough to `f.increase(3)`.

**Comment by damithc** [comment]
> I realise that the code in part c does not have `break` statements in the `switch-case` block. Will this affect the sequence diagram? Because if `status == 2`, the code will execute `f.increase(2)`, then it will fallthrough to `f.increase(3)`.

Good catch @aexolate The omission of the `break;` was unintentional. I've uploaded a new version with `break;` added.

**Comment by A1WAYSD** [comment]
@damithc Hi Prof, may I ask 
1. In (a) can we show the dependency from Section to Handout even if there's an association between them? Since textbook mentioned "we need not show that as a dependency arrow in the class diagram if the association is already indicated in the diagram", but here the navigability is from Handout to Section so drawing the dash line from Section to Handout does add extra value?
2. Can we show dependency in Object diagram?
3. In (c) where target object of run() is not specified, how can we name the target object?

Thank you!

**Comment by kiwibang** [comment]
> For the activity diagram question, I agree with @TeeRenJing 's implementation of the if-else portion. I think it makes more sense to have 3 different cases:
> 
> 1. Where p1, p2 and p3 are all positive,
> 2. There are 3 consecutive failures, and
> 3. Else
> 
> However, I also added a box containing "user prompted to place hand again" to the else line. I feel like @kiwibang 's way of representing the if-else statements is inaccurate. According to his diagram, 3 failures causes the user to be prompted to place their hand again, followed by the door being disabled, which doesn't make sense.
> 

Yep, agreed. Mine seems to be rather convoluted and perhaps wrong. 

> With regards to the sequence diagram, within the `[alt]` block, I believe an empty `[else]` block is necessary. To my understanding, inside the `[alt]` block, exactly one scenario has to be chosen.

Agreed, I feel that there should be an empty else block to handle the scenarios where status is neither 2 nor 3.

Thanks for the valuable inputs! 👍

**Comment by kiwibang** [comment]
@ktzy0305 I have some questions regarding some parts of your diagram. Could you help to clarify them? TIA!

(a) 
- For the association between `Brochure` and `Handout`, should it be 2 arrows instead? I feel that this association is not bi-directional but rather 2 uni-directional associations. Not very sure though
- For the dependency from `Section` to `Handout`, do we need to draw this? The notes say that if there is already an association between the two classes, then the dependency arrow is redundant and thus can be omitted. However, the association between the `Section` to `Handout` is in a different direction. In such a case, do we have to draw the dependency arrow like your diagram and mine, or can we omit it? 

(b)
- Do we need to show the composition relationship in an object diagram? Or can that be omitted? 
- Is there a strict rule as to whether we can give names to unamed objects?

(c) 
- Is there supposed to be a self-invocation in the activation bar for your `pad` method? Since the `pad` method calls the `increase` method, both of which are methods of the same `Factory` class

(d) 
- Perhaps the corners of each activity can be rounded more? It seems to be a little pointy

Other than that, I see no issues! These are really nicely drawn diagrams that I could never hope to achieve! 👍

**Comment by damithc** [comment]
> * In (a) can we show the dependency from Section to Handout even if there's an association between them? Since textbook mentioned "we need not show that as a dependency arrow in the class diagram if the association is already indicated in the diagram", but here the navigability is from Handout to Section so drawing the dash line from Section to Handout does add extra value?

@A1WAYSD The association is in the other direction. So, it doesn't imply this dependency. 

> * Can we show dependency in Object diagram?

Dependencies exist between classes, not objects.

> * In (c) where target object of run() is not specified, how can we name the target object?

You can give it any name or keep it empty. It will not appear in answer options in part 3 of the exam.

**Comment by damithc** [comment]
> * Do we need to show the composition relationship in an object diagram? Or can that be omitted?

@kiwibang good to show, as it adds value.

> * Is there a strict rule as to whether we can give names to unamed objects?

No. You can give names if it adds value e.g., if you need to refer to each object from elsewhere.

**Comment by damithc** [comment]
@ktzy0305 there's a minor notation error in the below:
![image](https://github.com/nus-cs2103-AY2324S1/forum/assets/1673303/d101a958-8c1b-4607-9f7b-5b7e1325f1e5)

**Comment by A1WAYSD** [comment]
@damithc Thanks Prof for bringing up that dependencies only exist between classes:D

**Comment by ktzy0305** [comment]
@damithc Thanks prof! Didn't realise I drew an extra set of curly braces.

> @ktzy0305 there's a minor notation error in the below: ![image](https://private-user-images.githubusercontent.com/1673303/286857144-d101a958-8c1b-4607-9f7b-5b7e1325f1e5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEzMzE0MDUsIm5iZiI6MTcwMTMzMTEwNSwicGF0aCI6Ii8xNjczMzAzLzI4Njg1NzE0NC1kMTAxYTk1OC04YzFiLTQ2MDctOWY3Yi01YjdlMTMyNWYxZTUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTEzMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzExMzBUMDc1ODI1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTM3MjdmNzc4MmEwYTg3ZDAzMDMxZTY3YTc4Nzc2ZTViYzFhNjFlZGFiYzE2ZWYyNmE2YzgwMWZjYmMzM2U2NiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.ygjqYSyMH9AkM-ZZL3N3kpzwGZ4TYtP1EXAHvLj5Ig0)

**Comment by ktzy0305** [comment]
Thank you prof @damithc for clarifying!! 

> (a)
> * For the association between `Brochure` and `Handout`, should it be 2 arrows instead? I feel that this association is not bi-directional but rather 2 uni-directional associations. Not very sure though

I am not very sure of the context of this question. It would seem like having two uni-directional associations would be correct too. If the question mentioned that when a `Brochure` is referenced by a `Handout` object, the `Brochure` object's cover page is not necessarily the same `Handout`, then it would be definitely 2 uni-directional associations.


> (c)
> 
> * Is there supposed to be a self-invocation in the activation bar for your `pad` method? Since the `pad` method calls the `increase` method, both of which are methods of the same `Factory` class

@kiwibang You're right, I am missing a self invocation for `increase(5)` in the `f:Factory` object.

**Comment by A1WAYSD** [comment]
> @Elijah5399 From the textbook, it is ok for none of the alt blocks to be taken, so I believe an empty else block is not necessary.
> 
> ![image](https://private-user-images.githubusercontent.com/122219138/286456480-2fe9f433-1175-42c4-a72d-3cd0d579edd9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEzNjIyMjcsIm5iZiI6MTcwMTM2MTkyNywicGF0aCI6Ii8xMjIyMTkxMzgvMjg2NDU2NDgwLTJmZTlmNDMzLTExNzUtNDJjNC1hNzJkLTNjZDBkNTc5ZWRkOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMTMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTEzMFQxNjMyMDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xY2NiYWYyMjlkZWRhYWFkNjBkMjFiNDJjN2MyZWEyZjg2NGYxNjU3NGU1OTQwYTFmMDlkM2QwYWY0ODljMjMzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.9-ehrblJPswLV2RHVeNBxDjt29RQ6_NIZbWvs306sDI)

@damithc Hi Prof, just to clarify for alt path in sequence diagram here, is it okay to skip all of the path as mentioned here, or one of the path in alt must be executed? Thank you!

**Comment by s-peiran** [comment]
I have some general questions

![image](https://github.com/nus-cs2103-AY2324S1/forum/assets/96849432/1682d09e-3c7e-45fb-9ece-265fb65b0a33)
if there are two arrows in the case above, would it matter which arrow i label the multiplicity at?

![image](https://github.com/nus-cs2103-AY2324S1/forum/assets/96849432/549ffb70-04fe-4d10-9c0a-4acd8f5745f0)
Would it be optional to write the return value here?

**Comment by damithc** [comment]
> ![image](https://private-user-images.githubusercontent.com/96849432/287000392-1682d09e-3c7e-45fb-9ece-265fb65b0a33.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE0NDQ0MjIsIm5iZiI6MTcwMTQ0NDEyMiwicGF0aCI6Ii85Njg0OTQzMi8yODcwMDAzOTItMTY4MmQwOWUtM2M3ZS00NWZiLTllY2UtMjY1ZmI2NWIwYTMzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjAxVDE1MjIwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE5YWI2ZWFlNWJhMDE4Zjg1NTU2NGI5MzAyYjEyOGQ3ODQ4NzQ4YzhmNjNhOTVjZmYyODQ5OWU0YjIzNzQ4MDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.stRLyuxabjyNUk9jKZQx-G6urJw4vy_nHJvj36ZbLfI)
> if there are two arrows in the case above, would it matter which arrow i label the multiplicity at?

@s-peiran Yes, it should be given at the line that represents the corresponding association.

> ![image](https://private-user-images.githubusercontent.com/96849432/287000817-549ffb70-04fe-4d10-9c0a-4acd8f5745f0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE0NDQ0MjIsIm5iZiI6MTcwMTQ0NDEyMiwicGF0aCI6Ii85Njg0OTQzMi8yODcwMDA4MTctNTQ5ZmZiNzAtMDRmZS00ZDEwLTljMGEtNGFjZDhmNTc0NWYwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjAxVDE1MjIwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgxZjQzZTFkZTMyNTNjNDIzYjk4YTQ1MDUxYjlmMGNjNWY4MTYxZTliYTAyNjNkOGU2YWQ1ZDRlZmM4NTU0ZmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.tKMJzlxIiuVYgd8eMYn1EqSDOFQulpZO4LB-zkr0wTo)
> Would it be optional to write the return value here?

Optional in the case of constructors (as it is already known the constructor returns the new object).
